### ISSUE TITLE
Redirects do not work after upgrade to 8.11

### ISSUE THREAD

**Issue by Jacob87** [umbraco/Umbraco-CMS#10066]
I had to run a query in the database after upgrading my Umbraco Cloud project from 8.9 to 8.11 to fix all the existing redirects.

## Umbraco version

I am seeing this issue on Umbraco version: 8.11.2

### Specifics

Recently I upgraded an Umbraco Cloud project from 8.9 --> 8.10.1 --> 8.11.1. After finishing the upgrade, I noticed that all redirects stopped working and would simply return 404 on the source URL. Upon further investigation, I noticed that redirects created *after* the upgrade had a value in the `culture` column in the database table for redirects (umbracoRedirectUrl) while existing redirects simply had a `NULL` value.

Running a query to update all of the old redirects to set the culture to my default (luckily I only had one language) seems to have fixed the issue:

```sql
UPDATE umbracoRedirectUrl SET culture = ‘en-us’ WHERE culture IS NULL
```

### Steps to reproduce

1. Create a redirect in redirect URL management
2. Open up the database and set the culture to NULL on the redirect
3. Redirect no longer works
4. Set the culture back to the original value (`en-us` for instance)
5. Redirect starts working

### Expected result

All existing redirects continually work after an upgrade.

### Actual result

Redirects stopped working, perhaps due to a missing migration.

**Comment by nul800sebastiaan** [comment]
Hmmm, that's not right, we'll definitely have to have a look, thanks for the detailed report! 👍

**Comment by nzdev** [comment]
Also doesn't work going from 7.14.0 -> 8.11.0 -> 8.12.2. I can confirm the workaround worked.

**Comment by nul800sebastiaan** [comment]
Hmm, I can't reproduce this problem at all. It would be helpful to have the before and after upgrade situation in order to see if this is still a problem in 8.13, but it doesn't look like it.

I tried with both a page that is not culture variant and a page that is culture variant. If you have a before database that doesn't work after upgrade then that would be great.

I will close this one for now, but if we have a good before & after case then we can have another look.

**Comment by Jacob87** [comment]
Hi @nul800sebastiaan 

You are saying you cannot reproduce it, so does that mean that your redirects migrated just fine following the upgrade path? 

What kind of before and after upgrade situation would you be interested in? Just redirects or nodes as well?

I realise perhaps I wasn't too clear on the situation on 8.11 and going forward, but here goes: Redirects created _before_ the upgrade did not work before I ran the query in the OP, but redirects created _after_ the upgrade worked just fine. In fact the solution was simply to just mirror the redirects from before so they looked like the redirects from after, if that makes sense. All that was missing was a value in the `culture` column.

**Comment by nul800sebastiaan** [comment]
> so does that mean that your redirects migrated just fine following the upgrade path?

Not exactly, I'm saying that I had redirects (both culture specific and not culture specific) and they worked just fine after upgrading to 8.13.0. I did look in the `culture` column as well, and `null` values worked for me.

**Comment by dinc5150** [comment]
I just had the same issue when migrating a site from V7.15 to V8.13.0

I've got a copy of the before / after of the database if you need some context of the data. 

Also found this forum post before I found this github issue, which seems to be due to the same issue - https://our.umbraco.com/forum/using-umbraco-and-getting-started/104868-umbraco-redirect-umbraco-url-management-not-working#comment-330206

**Comment by agaarp** [comment]
Chiming in here, I just did a migration from a V7.7 site to V8.14.0 and all of the redirects that came over from the V7.7 db have null in the culture column. Trying to hit a redirected link just gives me a 404. Updating the null values to 'en-us' fixed the issue.

**Comment by iOvergaard** [comment]
That is the exact issue I had as well, @agaarp. My project originally was migrated from Umbraco 7, and I guess it worked on Umbraco 8.9 with NULL value in that column, but something changed from 8.10 possibly, and the migration was apparently not updated accordingly.

**Comment by nathanwoulfe** [comment]
This is still an issue - we migrated from 7.15.5 to 8.16.2, redirects all stopped working. Updating culture to an empty string fixes it (site is invariant, so no need to set a culture).

**Comment by stevenstanley** [comment]
This still seems to be an issue, after upgrading to version 8.17.1 though when I look in the DB all the culture values are set to ''.

**Comment by RasmusUmbraco** [comment]
I had a user report this and help me reproduce it. The issue appears on a fresh v8 project as well.

These are the steps:
1. Create a new Umbraco 8 project on Cloud (8.18 in this case) with a Dev environment
2. Add some schema/doctypes to both environments.
3. Create a node
4. Push it to the next environment (Live)
5. Rename the node on dev - you will see that the automatic redirect for the old url works
6. Push the renamed node to live - the redirect is registered, but it doesn't work

The redirect is recognized on Live:
![image](https://user-images.githubusercontent.com/88326257/158978650-67a5e376-d663-4fd2-b6fb-b07bae873020.png)

But it does not work:
![image](https://user-images.githubusercontent.com/88326257/158978666-9ae1a020-1940-4e00-be97-878a6eff4a89.png)

umbracoRedirectUrl - Culture is blank on Dev
![image](https://user-images.githubusercontent.com/88326257/158978689-49899a2d-9771-4911-a76e-a73ae5eefe0d.png)

umbracoRedirectUrl – Culture is null on Live
![image](https://user-images.githubusercontent.com/88326257/158978719-7a462307-2420-4e14-8d06-c959fd475a78.png)

**Comment by hartviglarsen** [comment]
@nul800sebastiaan Is this something that can be looked at?

Potentially all/a lot of newer Umbraco 8 sites running on Cloud won't have working redirects if they are using the recommended approach in regards to content editing (work on staging - transfer to live).

**Comment by AndyButland** [comment]
Have had a little dig into this one.  As far as I can see the issue arises because redirects created without a culture use `null` as a default (see [here](https://github.com/umbraco/Umbraco-CMS/pull/9653)).

But when retrieving redirect URLs, `string.Empty` is used for the check for those not matching a culture (see [here](https://github.com/umbraco/Umbraco-CMS/blob/76ecb295cc27a714d3b77fbef3d3bd824c1a9601/src/Umbraco.Core/Persistence/Repositories/Implement/RedirectUrlRepository.cs#L171)).

We can amend Umbraco Deploy to provide an empty string as the default value for a culture.

But I think this probably needs a fix in the CMS, to amend the retrieval behaviour to look for redirects with a culture value of `null` _or_ `string.Empty`.  That way it doesn't matter if existing redirects use null or "", they'll still be retrieved.

**Comment by nul800sebastiaan** [comment]
The missing null check will be fixed in v8.18.1, which should fix the problem for non-Deploy users. I was able to reproduce it now and fix it in https://github.com/umbraco/Umbraco-CMS/pull/12154

The same fix will be applied for 9.4.0.

For Deploy users on lower versions of Umbraco, there will be a fix in Deploy as well, which won't set the culture to null but to `string.Empty` instead.
